### PR TITLE
feat: prefer older libssl versions rather than the newest ones (partial revert of #17126)

### LINKED ISSUE #17264: #17126 broke the `platform-serverless-vercel/vercel-cli` ecosystem tests
- [x] we should understand why that's the case
- [x] revert the PR and gradually add back the `openssl` improvements we intended, by monitoring the ecosystem-tests

### PR DESCRIPTION
**Pull request by jkomyno**
Integration PR (which deprecates https://github.com/prisma/prisma/pull/17240) that reverts the "pick up the most recent libssl.so version" feature from prisma/prisma#17126.

Closes https://github.com/prisma/prisma/issues/17264.

## Vercel issues

#17240 broke ecosystem-tests for Vercel serverless.
As it turns out, Vercel uses a different environment for building and running serverless functions, and these environments have different `libssl.so` system dependencies installed.

The Vercel **build** environment has the following `libssl.so` versions installed, as can be seen via `ls -r /lib64 | grep libssl.so`:

```sh
libssl.so.1.1.1g
libssl.so.1.1
libssl.so.1.0.2k
libssl.so.10
libssl.so
```

which resolves to the `openssl-1.1.x` binaryTarget.

The Vercel **runtime**, on the other hand, has only:

```sh
libssl.so.1.0.2k
libssl.so.10
```

which resolves to `openssl-1.0.x` binaryTarget.

Prisma Client would then fail because it cannot find `libquery-openssl-1.1.x` at runtime.

Takeaway: for the time being we shouldn't use the most recent versions of `libssl.so`, rather the oldest ones. It's ugly :(

### CODE REVIEW

**Comment by janpio** [comment]
Ecosystem tests run that show that this indeed fixes the failing vercel ecosystem tests: https://github.com/prisma/ecosystem-tests/actions/runs/3892460754 (The remaining deno failure is known and expected)

File: packages/get-platform/src/getPlatform.ts [hunk 1]
```diff
@@ -156,34 +156,33 @@ type GetOpenSSLVersionParams = {
...
-    .with({ distro: 'rhel' }, async () => {
-      /* Linux Red Hat, OpenSuse etc */
-      debug('Trying platform-specific paths for "rhel"')
-      return Promise.resolve(['/lib64', '/usr/lib64'])
-    })
+    // .with({ distro: 'rhel' }, () => {
+    //   /* Linux Red Hat, OpenSuse etc */
+    //   debug('Trying platform-specific paths for "rhel"')
+    //   return ['/lib64', '/usr/lib64']
+    // })
```
Comment (jkomyno) [inline packages/get-platform/src/getPlatform.ts:178]:
```suggestion
    .with({ distro: 'rhel' }, () => {
      /* Linux Red Hat, OpenSuse etc */
      debug('Trying platform-specific paths for "rhel"')
      return ['/lib64', '/usr/lib64']
    })
```

**Comment by aqrln** [comment]
> The Vercel **build** environment has the following `libssl.so` versions installed, as can be seen via `ls -r /lib64 | grep libssl.so`:
> 
> ```sh
> libssl.so.1.1.1g
> libssl.so.1.1
> libssl.so.1.0.2k
> libssl.so.10
> libssl.so
> ```
> 
> which resolves to the `openssl-1.0.x` binaryTarget.
> 
> The Vercel **runtime**, on the other hand, has only:
> 
> ```sh
> libssl.so.1.0.2k
> libssl.so.10
> ```
> 
> which resolves to `openssl-1.1.x` binaryTarget.

Should this be the other way around, i.e. 1.1 in the first case and 1.0 in the second case? (I assume it's a typo, but wanted to confirm just in case.)

**Comment by jkomyno** [comment]
- ecosystems-tests run: https://github.com/prisma/ecosystem-tests/actions/runs/3903136873/jobs/6667102884
- `docker/alpine:3.17` is failing merely because it’s expecting `openssl-3.0.x` rather than `openssl-1.1.x` (both versions of libssl are installed in the system, and we’re selecting the oldest one)
- many `platforms-serverless-vercel` tests that were failing before. now pass
- however, `vercel-cli` still fails because we detect that `Postinstall hook was NOT ADDED`, in which case we explicitly exit with error code `1`.

**Comment by jkomyno** [comment]
Incidentally, this PR may close https://github.com/prisma/prisma/issues/14042. See [this comment](https://github.com/prisma/prisma/issues/14042#issuecomment-1380692643).

**Comment by jkomyno** [comment]
## TODOs (if this PR is accepted as is)

- [ ] update `EXPECTED_PRISMA_TARGET_PLATFORM` for `docker/alpine:3.17` images to expect `openssl-1.1.x` rather than `openssl-3.0.x`
- [ ] update `platform-serverless gcp/firebase` to expect `openssl-1.0.x` rather than `openssl-1.1.x`

File: packages/get-platform/src/getPlatform.ts [hunk 2]
```diff
@@ -156,34 +156,33 @@ type GetOpenSSLVersionParams = {
  * On Linux, returns the OpenSSL version excluding the patch version, e.g. "1.1.x".
  * Reading the version from the libssl.so file is more reliable than reading it from the openssl binary.
  * This function never throws.
- *
- * TODO: we should probably validate the output of this function, and ensure it's contained in `supportedLibSSLVersions`
  */
 export async function getSSLVersion(args: GetOpenSSLVersionParams): Promise<GetOSResult['libssl'] | undefined> {
-  const libsslSpecificPaths = await match(args)
-    .with({ distro: 'musl' }, async () => {
+  const archFromUname = await getArchFromUname()
```
Comment (aqrln) [inline packages/get-platform/src/getPlatform.ts:161]: Can we move it to where it's used so we don't call `uname` on platforms where we don't need it?

File: packages/get-platform/src/getPlatform.ts [hunk 3]
```diff
@@ -156,34 +156,33 @@ type GetOpenSSLVersionParams = {
...
       /* Linux Red Hat, OpenSuse etc */
       debug('Trying platform-specific paths for "rhel"')
-      return Promise.resolve(['/lib64', '/usr/lib64'])
+      return ['/lib64', '/usr/lib64']
     })
     .otherwise(({ distro, arch }) => {
       /* Other Linux distros, we don't do anything specific and fall back to the next blocks */
       debug(`Don't know any platform-specific paths for "${distro}" on ${arch}`)
-      return Promise.resolve(undefined)
+      return undefined
```
Comment (aqrln) [inline packages/get-platform/src/getPlatform.ts:182]: Would it be better to return an empty array here and remove the default value below on line 185?

File: packages/get-platform/src/getPlatform.ts [hunk 4]
```diff
@@ -197,9 +196,9 @@ export async function getSSLVersion(args: GetOpenSSLVersionParams): Promise<GetO
 
   debug('Falling back to "ldconfig" and other generic paths')
   const libsslFilename: string | undefined = await getFirstSuccessfulExec([
-    'ldconfig -p | sed "s/.*=>s*//" | sed "s/.*///" | grep ssl | sort -r',
-    'ls -r /lib64 | grep ssl',
-    'ls -r /usr/lib64 | grep ssl',
+    'ldconfig -p | sed "s/.*=>s*//" | sed "s/.*///" | grep ssl',
```
Comment (aqrln) [inline packages/get-platform/src/getPlatform.ts:199]: maybe

```suggestion
    'ldconfig -p | sed "s/.*=>s*//" | sed "s/.*///" | grep ssl | sort',
```

to be safe

Comment (aqrln) [reply packages/get-platform/src/getPlatform.ts:199]: or is it guaranteed to be lexicographically ordered already?

Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts:199]: Hi, I also had second thoughts on this. It's probably better to keep `| sort` and be sure with it, as the paths returned by `ldconfig -p` may have wildly different basedirs

File: packages/get-platform/src/getPlatform.ts (hunk 3 above)
Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts:182]: Yes, we can do that

File: packages/get-platform/src/getPlatform.ts (hunk 2 above)
Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts:161]: I'd rather not: this was something we did before, but made it cumbersome dealing with all the non-`debian` pattern branches: they had useless `async` declarations and needed `Promise.resolve` before returning the result. Besides, The function that computes `uname` is guaranteed to not raise errors by design, is cheap to call and lies in a non-performance-critical block of Prisma

**Comment by Jolg42** [comment]
Recent ecosystem-test run https://github.com/prisma/ecosystem-tests/actions/runs/3910377877

**Review by aqrln — APPROVED** [review]

**Pull request merged**